core/adminserver/admin_server.go

```diff
@@ -167,12 +148,53 @@ func (s *Server) GetDomain(ctx context.Context, in *pb.GetDomainRequest) (*pb.Do
 	return info, nil
 }
 
+// privKeyOrGen returns the message inside privKey if privKey is not nil,
+// otherwise, it generates a new key with keygen.
+func privKeyOrGen(ctx context.Context, privKey *any.Any, keygen keys.ProtoGenerator) (proto.Message, error) {
+	if privKey != nil {
+		var keyProto ptypes.DynamicAny
+		if err := ptypes.UnmarshalAny(privKey, &keyProto); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal privatekey: %v", err)
+		}
+		return keyProto.Message, nil
+	}
+	return keygen(ctx, &keyspb.Specification{
```

Comment:
Maybe pull out the keyspec as it's used in several places.

Reply:
Done